Introduce Backups

stephenplusplus opened the pull request:
## API Changes

#### Access or create Backup objects through a Cluster reference
```js
cluster.backup(id) // references a pre-existing "Backup" object (no remote calls)
cluster.createBackup({ id: 'backup-id', table: 'table-id' }) // creates and returns a "Backup" object
cluster.getBackups() // returns "Backup" objects
cluster.getBackupsStream() // returns "Backup" objects as a readable stream
```

#### Access Backup objects through an Instance reference
```js
instance.createTableFromBackup({ table: 'table-id', backup: 'backup-id' }, gaxOptions)
instance.getBackups() // returns "Backup" objects
instance.getBackupsStream() // returns "Backup" objects as a readable stream
```

#### Create a Backup through a Table reference
```js
table.createBackup(id, metadata) // creates and returns a "Backup" object
```

#### Once you have a reference to a Backup object:
*All methods below besides "restore()" are common CRUD methods defined throughout our API*
```js
backup.create({ table: 'table-id', metadata: {...} })
backup.delete()
backup.get()
backup.getMetadata()
backup.restore('table-id')
backup.setMetadata({...})
```

#### References

- [Original implementation](https://github.com/googleapis/nodejs-bigtable/pull/761)
- [Proto file](https://github.com/googleapis/googleapis/blob/master/google/bigtable/admin/v2/bigtable_table_admin.proto)

## To Dos

- [x] Source code
- [x] TypeScript annotations
- [x] JSDoc parameters
- [x] Unit tests
  - [x] Backup
  - [x] Cluster
  - [x] Index
  - [x] Instance
  - [x] Table
- [x] System tests

This PR splits the work from #761 into three parts-- source code, tests, and then documentation and samples to come in a follow-up PR. This should make reviewing easier, as well as lets us lock in the source code changes before attempting to adapt the previously written tests and samples.

Currently, this PR has only the source code. Most of it was carried from #761, but there were some structural changes to follow our conventions:

```diff
- Backup(id, cluster, fields)
+ Backup(cluster, id)

get backup.expireDate()
get backup.startDate()
get backup.endDate()

- backup.create(table, metadata)
+ backup.create({table, metadata})
backup.delete()
backup.get()
+ backup.getMetadata()
backup.restore(tableId, gaxOptions)
- backup.update(metadata, gaxOptions)
+ backup.setMetadata(metadata, gaxOptions)

+ cluster.backup(id)
+ cluster.createBackup({id, table}, gaxOptions)
cluster.getBackups(options)
cluster.getBackupsStream(options)

- instance.backup(id, cluster, fields)
- instance.createTableFromBackup(tableId, backup, gaxOptions)
+ instance.createTableFromBackup({table, backup}, gaxOptions)
+ instance.getBackups(options, callback)
+ instance.getBackupsStream(options, callback)

- table.backup(id, metadata, gaxOptions)
+ table.createBackup(id, metadata, gaxOptions)
```

@AVaksman could you please review this when you get a chance? I will likely need special attention to the TypeScript, making sure I'm properly compliant with our conventions and use cases I may have missed. Thank you!

omniadonum committed:
feat(backups): implement create/get/list/delete backups with sample

omniadonum committed:
docs(backups): add examples, samples, extra method signatures, fix delete timeouts, expireTime is required

omniadonum committed:
feat(table): add backup helper right on the table that auto selects the cluster

omniadonum committed:
feat(backups): implement update and restore backup methods, refactor expire to field encapsulation for flexibility

omniadonum committed:
feat(backups): add Backup class wrapper with helper methods, include precise-date for backup timestamps

omniadonum committed:
feat(backups): stream list of backups, export new backup types, remove dupe proto import, import reorder

omniadonum committed:
fix(backups): support optimize replstate during backup, support Date for backup timestamps, show Backup usage in sample

omniadonum committed:
fix(table): after refactoring tests to promisify fixed missing id check bug

omniadonum committed:
fix(backups): id checks for cluster, keep cluster promise

omniadonum committed:
test(backups): table backup helper

omniadonum committed:
test: keep promisify during tests, fix improper tests that did not consider promise wrapping

omniadonum committed:
fix: after refactoring tests to promisify fixed missing id check bugs

omniadonum committed:
fix(backups): list pagination full impl, expiretime checks, allow table name instead of reference, orig backup as public metadata

omniadonum committed:
test(backups): cluster, instance, and table coverage

omniadonum committed:
test(backups): full unit test coverage for all new backup code

omniadonum committed:
test(backups): system tests for all backup methods, added asBackup casting helper and update sample accordingly

omniadonum committed:
docs(backups): fix sample test case stdout partial match

omniadonum committed:
feat(backups): Backup is the primary class representing and providing to interact with backups

stephenplusplus committed:
code changes

stephenplusplus committed:
temporarily remove tests and samples

stephenplusplus committed:
remove test and revert copyright

stephenplusplus committed:
lint

stephenplusplus committed:
revert change

stephenplusplus committed:
change pagination code

stephenplusplus committed:
ignore any

AVaksman commented on src/cluster.ts:
```diff
@@ -437,6 +655,12 @@ Please use the format 'my-cluster' or '${instance.name}/clusters/my-cluster'.`);
+ *
+ * These methods can be auto-paginated.
+ */
+paginator.extend(Cluster, ['getBackups']);
```
I believe this is no longer needed

AVaksman commented on src/instance.ts:
```diff
@@ -1084,6 +1092,63 @@ Please use the format 'my-instance' or '${bigtable.projectName}/instances/my-ins
+   *     restore of the form
+   *     `projects/<project>/instances/<instance>/clusters/<cluster>/backups/<backup>`,
+   *     or a Backup instance.
+   * @param {CallOptions | RestoreTableCallback} [config.gaxOptions] Request
```
```suggestion
   * @param {CallOptions} [config.gaxOptions] Request
```

AVaksman commented on src/instance.ts:
```diff
@@ -1084,6 +1092,63 @@ Please use the format 'my-instance' or '${bigtable.projectName}/instances/my-ins
+      throw new Error('An table id is required to restore from a backup.');
+    }
+
+    const tableAdminClient = this.bigtable.api[
```
`this.bigtable.api` may be still empty

AVaksman commented on src/instance.ts:
```diff
@@ -1084,6 +1092,63 @@ Please use the format 'my-instance' or '${bigtable.projectName}/instances/my-ins
+    if (config.backup instanceof Backup) {
+      backup = config.backup;
+    } else {
+      const clusterId = tableAdminClient
```
`tableAdminClient` may be `undefined`

AVaksman commented on src/cluster.ts:
```diff
@@ -321,6 +429,116 @@ Please use the format 'my-cluster' or '${instance.name}/clusters/my-cluster'.`);
+   *     this.end();
+   *   });
+   */
+  getBackupsStream(options: GetBackupsOptions): NodeJS.ReadableStream {
```
For this method to work `listBackupsStream: true` needs to be added to [`gapicStreamingMethods`](https://github.com/googleapis/nodejs-bigtable/blob/2a199aae5646fe4e19836e13a61fcb83b527e4c5/src/index.ts#L753)

stephenplusplus committed:
TS help round 1

stephenplusplus commented on src/backup.ts:
```diff
@@ -0,0 +1,505 @@
+        method: 'restoreTable',
+        reqOpts: {
+          parent: this.cluster.name,
+          tableId: table.name,
```
Woops, thanks!

stephenplusplus committed:
simplify name checking

stephenplusplus committed:
remove raw gapic api use

stephenplusplus committed:
simplify cluster.createBackup() and backup.create()

stephenplusplus committed:
elaborate jsdocs, correct create() usage

stephenplusplus commented on src/backup.ts:
```diff
@@ -0,0 +1,505 @@
+      typeof gaxOptionsOrCallback === 'function' ? gaxOptionsOrCallback : cb!;
+
+    // @TODO `this.metadata` isn't a `Backup`. Figure out why it has to be cast.
+    this.get(gaxOpts, err => callback(err, this.metadata as Backup));
```
I didn't see the suggestion about moving the function body from `getMetadata()` to `get()`. I'm not opposed to it, but why is one better than the other? (I've since made the other corrections to these functions, and they all seem to be co-existing happily)

AVaksman replied:
Perhaps didn't click "Add review comment"
My suggestion was to move the function body from `get()` to `getMetadata()` as the snippet above.
Then in `get()` do 
```ts
    this.getMetadata(
      gaxOptions,
      (err, metadata) => {
        callback(err, err ? null : this, metadata);
      }
    );
```
Following an existing example in [Cluster](https://github.com/googleapis/nodejs-bigtable/blob/cb77d569551bab8dd2897c31ea5a47aa0ae99706/src/cluster.ts#L305-L322), [Instance](https://github.com/googleapis/nodejs-bigtable/blob/cb77d569551bab8dd2897c31ea5a47aa0ae99706/src/instance.ts#L620-L635)

stephenplusplus replied:
Gotcha, thanks!

stephenplusplus committed:
backup name formatting and error handling

stephenplusplus commented:
@AVaksman thank you very much for the thorough review! I've made the changes (thank you for all of the suggestions) and left just one thing: https://github.com/googleapis/nodejs-bigtable/pull/794#discussion_r482692426.

JustinBeckwith commented on src/table.ts:
```diff
@@ -512,6 +521,74 @@ Please use the format 'prezzy' or '${instance.name}/tables/prezzy'.`);
+      throw new TypeError('An id is required to create a backup.');
+    }
+
+    this.getReplicationStates(config.gaxOptions)
```
The mixing of promises and callbacks is hurting my head a little here.  Is it like this because `getReplicationStates` doesn't support callbacks?

stephenplusplus committed:
remove paginator dep

stephenplusplus commented on src/cluster.ts:
```diff
@@ -321,6 +429,119 @@ Please use the format 'my-cluster' or '${instance.name}/clusters/my-cluster'.`);
+        gaxOpts: options.gaxOptions,
+      },
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      (err, ...resp: any[]) => {
```
@AVaksman could you advise here? This was taken from your suggestion on the other PR. I'm fighting a lot of TypeScript errors when I tamper with this.

AVaksman replied:
I took a stab at it as a part of an existing PR #731 where I am differentiating `bigtable.request` signature between `NormalCallback`, `LROCallback` and `PaginatedCallback`
```ts
export type RequestCallback<T = void, R = void> = T extends void
  ? LROCallback
  : R extends void
  ? NormalCallback<T>
  : PagedCallback<T, R>;

export interface NormalCallback<TResponse> {
  (err: ServiceError | null, res?: TResponse): void;
}

export interface PagedCallback<Item, Response> {
  (
    err: ServiceError | null,
    results?: Item[] | null,
    nextQuery?: {} | null,
    response?: Response | null
  ): void;
}

export interface LROCallback {
  (
    err: ServiceError | null,
    operation?: Operation,
    apiResponse?: IOperation
  ): void;
}
```

Then here the something like this would work
```ts
this.bigtable.request<
      google.bigtable.admin.v2.IBackup,
      google.bigtable.admin.v2.IListBackupsResponse
    >(
      {
        client: 'BigtableInstanceAdminClient',
        method: 'listAppProfiles',
        reqOpts,
        gaxOpts,
      },
      (err, IBackups, nextPageRequest, apiResponse) => {
        if (err) {
          callback(err);
          return;
        }
        const backups = IBackups!.map(backup => {
          const backupInstance = this.backup(backup.name!.split('/').pop()!);
          backupInstance.metadata = backup;
          return backupInstance;
        });
        const nextQuery = nextPageRequest!
          ? extend({}, options, nextPageRequest!)
          : null;
        callback(null, appProfiles, nextQuery, apiResponse!);
      }
    );
```
Is this an overkill?
`bigtable.request` signature changes could be done in a separate PR in a non breaking way.
WDYT?

AVaksman replied:
If agree, I can do the `bigtable.request` PR?

stephenplusplus replied:
Curious for @JustinBeckwith's thoughts. Whatever makes sense to you guys :)

For now, should I leave the code as-is? Also, please LMK if everything else looks correct so far and I'll get the tests going.

JustinBeckwith replied:
I am ok with leaving this as is for now, and coming back around to improve the under types here later.  I'm ok with it because I don't think they affect the public types.  Not a big deal.

stephenplusplus committed:
unpromisify

stephenplusplus commented:
Looks like you can't delete old @google-cla posts-- it'll just post it again 🤣

AVaksman commented:
I've seen commenting "@googlebot I consent." have helped @stephenplusplus

stephenplusplus committed:
switch logic between get/getMetadata to match other files

AVaksman commented on src/backup.ts:
```diff
@@ -0,0 +1,516 @@
+    const callback =
+      typeof gaxOptionsOrCallback === 'function' ? gaxOptionsOrCallback : cb!;
+
+    const {expireTime, ...restMetadata} = metadata;
```
As of right now the only updatable field is `expireTime`.
In general, do we want to require that "at least one field value should be provided" and throw if none provided?

stephenplusplus replied:
I don't feel too strongly, although I don't think we have a consistent precedent for that. I lean towards letting the API fail if it has an issue, when we don't have any code that requires it be defined (e.g. if we did "expireTime.toStruct()" without making sure if it works first).

AVaksman commented on src/cluster.ts:
```diff
@@ -201,6 +230,85 @@ Please use the format 'my-cluster' or '${instance.name}/clusters/my-cluster'.`);
+   *     https://googleapis.github.io/gax-nodejs/CallSettings.html.
+   * @param {CreateBackupCallback} [callback] The callback function.
+   * @param {?error} callback.err An error returned while making this request.
+   * @param {Cluster} callback.backup The newly created Backup.
```
```suggestion
   * @param {Backup} callback.backup The newly created Backup.
```

AVaksman commented on src/cluster.ts:
```diff
@@ -171,6 +190,16 @@ Please use the format 'my-cluster' or '${instance.name}/clusters/my-cluster'.`);
+   * @param {string} id The backup name or id.
+   * @returns {Backup}
+   */
+  backup(id: string): Backup {
```
`'backup'` should be added to promisifyAll exclude.

AVaksman commented on src/instance.ts:
```diff
@@ -1084,6 +1091,58 @@ Please use the format 'my-instance' or '${bigtable.projectName}/instances/my-ins
+        const clusterId = config.backup.match(/clusters\/([^/]+)/)![1];
+        backup = this.cluster(clusterId).backup(config.backup);
+      } catch (e) {
+        throw new Error('A complete backup name (path) is required.');
```
Or a backup object reference

AVaksman commented on src/instance.ts:
```diff
@@ -140,6 +141,12 @@ export type SetInstanceMetadataCallback = (
 export type SetInstanceMetadataResponse = [google.protobuf.Empty];
 
+export interface CreateTableFromBackupConfig {
+  table: string;
```
Should we split `tableId` param from the rest of the config values to have similar feel as `Instance.createTable()`?

stephenplusplus replied:
That is actually how it used to be, and I changed it to the config object. I didn't think about matching "createTable(tableId)", although that makes sense. However, I still think it's best to do a configuration object, as my reasoning was it's not intuitive from the method name which argument should come first, "table" or "backup". Each would be logical and could trip users up:

```js
await createTableFromBackup('my-backup', 'my-table') // :(
await createTableFromBackup('my-table', 'my-backup') // :(
await createTableFromBackup({table: 'my-table', backup: 'my-backup'}) // :)
```

LMK if you still think we should switch!

AVaksman replied:
Let's have the config object 👍

AVaksman commented on src/instance.ts:
```diff
@@ -1084,6 +1091,58 @@ Please use the format 'my-instance' or '${bigtable.projectName}/instances/my-ins
   }
 
+  createTableFromBackup(
+    config: CreateTableFromBackupConfig
```
Question about passing `tableId` separately from the rest of the config to have a similar feel as `Instance.createTable()`

AVaksman commented on src/table.ts:
```diff
@@ -512,6 +521,80 @@ Please use the format 'prezzy' or '${instance.name}/tables/prezzy'.`);
+   * Backups of tables originate from a specific cluster. This is a helper
+   * around `Cluster.createBackup` that automatically selects the first ready
+   * cluster from which a backup can be performed.
+   *
```
Should probably warn that this method uses another RPC?

stephenplusplus commented on src/backup.ts:
```diff
@@ -0,0 +1,516 @@
+
+export type SetMetadataCallback = (
+  err: ServiceError | null,
+  metadata: IBackup,
+  resp: IBackup
```
I tried to copy what ServiceObject#setMetadata does: https://github.com/googleapis/nodejs-common/blob/981a17650007d5c2a8b30051909baf2cc795d05d/src/service-object.ts#L510:

```js
callback!(err, this.metadata, res);
```

AVaksman replied:
Actually all other setMetadata functions return LROs.
So no preference here.

stephenplusplus committed:
pr feedback!

stephenplusplus commented on src/cluster.ts:
```diff
@@ -321,6 +429,119 @@ Please use the format 'my-cluster' or '${instance.name}/clusters/my-cluster'.`);
+    const reqOpts: google.bigtable.admin.v2.IListBackupsRequest = {
+      ...options,
+      parent: this.name,
+    };
```
Done! Although, please take a look as I had to make some changes to the way you had it written.

stephenplusplus committed:
correct JSdocs and simplify setMetadata

stephenplusplus committed:
rename types

stephenplusplus committed:
update promisify to support accessor methods

stephenplusplus committed:
test: intance.ts

stephenplusplus committed:
test: index.ts

stephenplusplus committed:
alphabetize

stephenplusplus committed:
test: cluster.ts

stephenplusplus committed:
Revert "alphabetize"

This reverts commit 4c1b46be4973033066ce0935bba7b06a57922b4c.

stephenplusplus committed:
test: table.ts

stephenplusplus committed:
test: backup.ts

stephenplusplus commented:
@AVaksman I believe I have completed the unit tests, although I'm always left wondering if there's one too many "any"s, or if I flubbed some other TypeScript best practice. If you could take a look, I would appreciate any help! (The 'test/backup.ts' file is especially any-heavy)

AVaksman commented on test/backup.ts:
```diff
@@ -0,0 +1,550 @@
+
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      (backup as any).bigtable.request = (config: any) => {
+        assert.deepStrictEqual(config.gaxOpts, gaxOptions);
```
```suggestion
        assert.strictEqual(config.gaxOpts, gaxOptions);
```

AVaksman commented on test/backup.ts:
```diff
@@ -0,0 +1,550 @@
+  let BACKUP_NAME: string;
+
+  let Backup: typeof backupTypes.Backup;
+  let backup: backupTypes.Backup;
```
It seems if declare `backup` type `any` here, most subsequent casts to any (`(backup as any).someMethod = ...`) could be avoided?

stephenplusplus replied:
Nice, thanks! (Woops)

AVaksman commented:
Are we interested in providing `instance#getBackups()` function to list all backups belonging to this instance across all clusters?

By providing `reqOpts.parent` in the form of `'projects/{project}/instances/{instance}/clusters/-'` 

WDYT?

stephenplusplus committed:
remove as anys

stephenplusplus committed:
pr review changes

stephenplusplus commented on test/table.ts:
```diff
@@ -234,6 +234,135 @@ describe('Bigtable/Table', () => {
+      };
+
+      table.getReplicationStates = (gaxOptions: {}, callback: Function) => {
+        callback(null, REPLICATION_STATES);
```
Nice catch!

stephenplusplus committed:
strict equal check

stephenplusplus committed:
add test

stephenplusplus committed:
pop the name

stephenplusplus committed:
add system tests & correct restore API call

stephenplusplus commented:
> Are we interested in providing `instance#getBackups()` function to list all backups belonging to this instance across all clusters?

Great idea, I missed that was possible. I'll add that in now.

Thank you for the review. I've addressed the other comments and added system tests.

stephenplusplus committed:
add instance#getBackups and instance#getBackupsStream

stephenplusplus commented:
`Instance#getBackups()` and `Instance#getBackupsStream()` added.

AVaksman commented on system-test/bigtable.ts:
```diff
@@ -1111,8 +1132,148 @@ describe('Bigtable', () => {
+
+      const expectedTime = expireTime.toStruct();
+      assert.deepStrictEqual(
+        BACKUP.metadata?.expireTime?.seconds?.toString(),
```
Do you want to use `BACKUP#expireDate()` here and in other tests?

stephenplusplus commented on system-test/bigtable.ts:
```diff
@@ -35,20 +36,39 @@ describe('Bigtable', () => {
-        return !timeCreated || timeCreated <= oneHourAgo;
-      });
+    const testInstances = instances.filter(i => i.id.match(PREFIX));
+    // .filter(i => {
+    //   const timeCreated = (i.metadata!.labels!.time_created as {}) as Date;
+    //   // Only delete stale resources.
+    //   const oneHourAgo = new Date(Date.now() - 3600000);
+    //   return !timeCreated || timeCreated <= oneHourAgo;
+    // });
```
Yes, oops. Didn’t mean to leave this in.

stephenplusplus committed:
address pr feedback

stephenplusplus commented:
@AVaksman thanks again. Good call using the new Instance#getBackups() method from the tests. Caught an issue where I wasn't replacing the "-" cluster ID with the actual one returned from the API.

stephenplusplus commented:
@AVaksman if all source and test code looks good, we can merge this in, then I'll be taking on the samples in a follow-up PR.

AVaksman commented on system-test/bigtable.ts:
```diff
@@ -1111,8 +1130,152 @@ describe('Bigtable', () => {
+    // testing and the naming convention used here does not actually influence
+    // the real functionality - it is just a way to keep things organized!
+    const backupIdFromCluster = generateId('backup');
+    const backupNameFromCluster = `${CLUSTER.name}/backups/${backupIdFromCluster}`;
```
Does it pay to use `replaceProjectId` or [`replaceProjectIdToken`](https://github.com/stephenplusplus/nodejs-bigtable/blob/spp--backups/src/index.ts#L15) here and a few lines below for backupNameFromTable?

AVaksman commented:
> if all source and test code looks good, we can merge this in, then I'll be taking on the samples in a follow-up PR.

Left a couple of suggestions. Can you tale a look? WDYT?

AVaksman commented on src/instance.ts:
```diff
@@ -827,12 +827,30 @@ Please use the format 'my-instance' or '${bigtable.projectName}/instances/my-ins
+      }
+
+      const backupsFromCorrectCluster = backups?.map(backup => {
+        const matchedClusterId = backup.metadata?.name?.match(
+          /clusters\/([^/]+)/
+        )![1];
+        const backupInstance = this.cluster(matchedClusterId!).backup(
+          backup.id
+        );
```
Or just 
```ts
this.instance.cluster(backup.name?.match(/clusters\/([^/]+)/)![1]!)
              .backup(backup.name!.split('/').pop()!)
```
without condition

stephenplusplus committed:
pr comments

stephenplusplus commented:
@AVaksman thank you again. Suggestions looked good to me 👍🏻

stephenplusplus commented on src/cluster.ts:
```diff
@@ -504,6 +504,7 @@ Please use the format 'my-cluster' or '${instance.name}/clusters/my-cluster'.`);
         if (resp[0]) {
           backups = resp[0].map((backup: IBackup) => {
             const backupInstance = this.backup(backup.name!.split('/').pop()!);
+            console.log('instance name when created', backupInstance.name);
```
🤦🏻

AVaksman commented:
I think it is good to go!

kolea2 commented on src/instance.ts:
```diff
@@ -29,6 +29,11 @@ import {
   GetAppProfilesCallback,
   GetAppProfilesResponse,
 } from './app-profile';
+import {
```
how come this is not on `table.ts`? In other clients, all backup operations are on the table admin client. ex: https://github.com/liubonan/java-bigtable/blob/master/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/admin/v2/BigtableTableAdminClient.java#L805. Let me know if my understanding is incorrect :)

stephenplusplus replied:
Hello! We ended up with these methods defined throughout various classes:

#### Access or create Backup objects through a Cluster reference
```js
cluster.backup(id) // references a pre-existing "Backup" object (no remote calls)
cluster.createBackup({ id: 'backup-id', table: 'table-id' }) // creates and returns a "Backup" object
cluster.getBackups() // returns "Backup" objects
cluster.getBackupsStream() // returns "Backup" objects as a readable stream
```

#### Access Backup objects through an Instance reference
```js
instance.createTableFromBackup({ table: 'table-id', backup: 'backup-id' }, gaxOptions)
instance.getBackups() // returns "Backup" objects
instance.getBackupsStream() // returns "Backup" objects as a readable stream
```

#### Create a Backup through a Table reference
```js
table.createBackup(id, metadata) // creates and returns a "Backup" object
```

#### Once you have a reference to a Backup object:
*All methods below besides "restore()" are common CRUD methods defined throughout our API*
```js
backup.create({ table: 'table-id', metadata: {...} })
backup.delete()
backup.get()
backup.getMetadata()
backup.restore('table-id')
backup.setMetadata({...})
```

AVaksman commented:
Might need a couple of unit tests to keep codecov happy

stephenplusplus committed:
address test coverage

JustinBeckwith commented on src/cluster.ts:
```diff
@@ -201,6 +239,91 @@ Please use the format 'my-cluster' or '${instance.name}/clusters/my-cluster'.`);
+      },
+    };
+
+    if (reqOpts.backup.expireTime instanceof Date) {
```
There used to be all kinds of nuance on when it was ok to use `instanceof Date`, and you had to do weird tricks to detect date objects.  It looks like `is` now does exactly what you're doing here though:
https://github.com/sindresorhus/is/blob/master/source/index.ts

So I'm good with it :)

JustinBeckwith commented on src/cluster.ts:
```diff
@@ -321,6 +444,150 @@ Please use the format 'my-cluster' or '${instance.name}/clusters/my-cluster'.`);
+    const callback =
+      typeof optionsOrCallback === 'function' ? optionsOrCallback : cb!;
+
+    options = extend(true, {}, options);
```
I spent a year trying to remove `extend` in favor of `Object.assign`, and I just want to say out loud that @stephenplusplus was right all along and I was wrong.

stephenplusplus replied:
Hahaha, stdlib got so close on this one!

stephenplusplus commented on src/backup.ts:
```diff
@@ -0,0 +1,542 @@
+  /**
+   * An expression for specifying the sort order of the results of the request.
+   * The string value should specify one or more fields in
+   * {@link google.bigtable.admin.v2.Backup|Backup}. The full syntax is
```
Hmm, well I don't actually know where that gets rendered. That's part of `interface GetBackupsOptions`, and I can't find where it went. It's not linked to where we use it from the `getBackups()` methods :\. All of the other links from descriptions seem to work, though!
